compilerProvider.js

```diff
@@ -0,0 +1,301 @@
+}
+
+/**
+ * Default configuration
+ * @type {Object}
+ */
+CompilerProvider.prototype.config = {
+  solc: null,
+  versionsUrl: 'https://solc-bin.ethereum.org/bin/list.json',
+  compilerUrl: 'https://solc-bin.ethereum.org/bin/',
+  compilerNpm: 'solc',
+  cache: true,
+}
+
+
+CompilerProvider.prototype.cachePath = findCacheDir({
+  name: 'truffle',
+  cwd: __dirname,
+  create: true,
+})
+
+//----------------------------------- Interface  ---------------------------------------------------
+
+/**
+ * Loads solc from four possible locations:
+ * - local node_modules            (param: <undefined>)
+ * - absolute path to a local solc (param: <path>)
+ * - a solc cache                  (param: <version-string> && cache contains version)
+ * - a remote solc-bin source      (param: <version-string> && version not cached)
+ *
+ * @param  {String} options [optional] options to pass to native solc binary
+ * @return {Module}          solc
+ */
+CompilerProvider.prototype.load = function(options){
+  const self = this;
+  const solc = self.config.solc;
+
+  return new Promise((accept, reject) => {
+    const useDefault = !solc;
+    const useLocal =   !useDefault && self.isLocal(solc);
+    const useRemote =  !useLocal;
+
+    if (useDefault) return accept(self.getDefault());
+    if (useLocal)   return accept(self.getLocal(solc));
+    if (useRemote)  return accept(self.getByUrl(solc));
+  });
+}
+
+/**
+ * Returns keys that can be used to specify which remote solc to fetch
+ * ```
+ *   latestRelease:  "0.4.21",
+ *   releases: ["0.4.21", ...],
+ *   prereleases: ["0.4.22-nightly.2018.4.10+commit.27385d6d", ...]
+ * ```
+ * @return {Object} See above
+ */
+CompilerProvider.prototype.getReleases = function(){
+  return this
+    .getVersions()
+    .then(list => {
+
+      // Prereleases
+      const prereleases = list
+          .builds
+          .filter(build => build['prerelease'])
+          .map(build => build['longVersion']);
+
+      // Releases
+      const releases = Object.keys(list.releases);
+
+      return {
+        prereleases: prereleases,
+        releases: releases,
+        latestRelease: list.latestRelease,
+      }
+    });
+}
+
+//------------------------------------ Getters -----------------------------------------------------
+
+/**
+ * Gets solc from local `node_modules`. Equivalent to `require("solc")`
+ * @return {Module} solc
+ */
+CompilerProvider.prototype.getDefault = function(){
+  // Temporary: Webpack has a problem w/ dynamic requires
+  //const compiler = require(this.config.compilerNpm);
+
+  const compiler = require('solc');
+  this.removeListener();
+  return compiler;
+}
+
+/**
+ * Gets an npm installed solc from specified absolute path.
+ * @param  {String} localPath
+ * @return {Module}
+ */
+CompilerProvider.prototype.getLocal = function(localPath){
+  const self = this;
+  let compiler;
+
+  try {
+    // Broken in built truffle: Webpack has a problem w/ dynamic requires
```

Comment:
wait, does this mean built truffle can't do this at all, or is there a workaround?

Reply:
Oh just realized this might be possible with `original-require` which is already externalized in the build [here](https://github.com/trufflesuite/truffle/blob/develop/cli.webpack.config.js#L42-L43).